### PR TITLE
Refactor IREP to work more like a build tool

### PR DESCRIPTION
**Pull request by tgamblin**
This PR refactors `irep` so that it can be used in a decoupled way in large codes.  Previously, the integration model for `irep` was to:

1. include `irep` in your code
2. add `wkt_*.h` files *to the `irep` installation itself*
3. build a `libIR.a` containing table definitions and all lookup structures, and
4. link your application with this `libIR.a` library.

This model means that every code needs to own its own `irep`, which makes `irep` hard to reuse across codes. This PR attempts to refactor the model so that `irep` is more easily consumed as a dependency of external projects.

## irep-generate

Previously, using all the code generations features of `irep` required Makefile magic and fairly in-depth knowledge of `irep`'s internals. This is all bundled into the `irep-generate` tool, which looks like this:

```
irep-generate [-h] [--mode PART] HEADERS

This script generates source code for IREP well-known-table (WKT)
libraries.

Generated code is printed to standard out.

Positional Arguments:
  HEADERS            wkt_*.h files from which to generate code, e.g.:
                       wkt_geometry.h wkt_physics.h ...

Generation options (by default, generate index)
  --mode index       generate only index
  --mode fortran     generate fortran module from a single wkt header
  --mode lua         generate loadable, nested lua tables
  --mode rst         generate restructured text (.rst) documentation

  --module-name      name for generated module (fortran mode only,
                     inferred from header name by default)

Documentation options (for use with --mode rst)
  --doc-dir DIR      documentation directory where we look for
                     details/intros for WKTs (default: .)

Environment variables:
  CPP                C preprocessor to use (default gcc -E)
  CPPFLAGS           flags for the C preprocessor

Help:
  -h, --help         display this message
```

`irep-generate` can be used to translate `wkt*.h` files into C, Fortran, Lua code, and it can also generate RST documentation. The particular mode of generation used is controlled by the `--mode` flag.
 
The most interesting options here are `--mode index` and `--mode fortran`. 

### Fortran generation

Running:

```
$ irep-generate --mode fortran wkt_foo.h
```

 will generate `wkt_foo.f`, which can be compiled into a Fortran module and `.o` file.  The `.o` file contains the definition for the table structure, and the `.mod` file is includable by other Fortran code.  You can use this type of code generation to generate WKT libraries for specific projects.

### Index generation

Running:

```
$ irep-generate --mode index wkt_foo.h wkt_bar.h wkt_baz.h
```

Will generate C code that IREP uses to look up the location of fields from well known tables.  This needs to be run on by each application using IREP, over all of the `wkt*.h` files included in that application.  If WKT files are not included.

### Lua code generation

Running:

```
$ irep-generate --mode lua wkt_foo.h
```

Will generate importable Lua code for the WKT module.

### Documentation generation

Running:

```
$ irep-generate --mode rst wkt_foo.h wkt_bar.h wkt_baz.h
```

Will generate RST documentation files for each well known table, as well as an `irep_types.rst` describing IREP data types.  You can include these RST files in [sphinx](https://www.sphinx-doc.org/) documentation for your code.

## Build integration

This version of irep contains build integration for `make` and `CMake`-based builds. 

* `make`-based builds can include the `wkt.mk` file included in this PR.  This defines implicit rules for generating `libfoo-wkt.a` and `libfoo-wkt-index.a` libraries.
* `CMake`-based builds can simply run `find_package(irep)` to auto-detect an `irep` installation.  This will define rules for building WKT libraries in CMake-based builds.

- [x] IREP build now creates a standalone `libIR.a` -- `wkt` libraries are separate.
- [x] add `irep-generate` with index, fortran, rst, and lua generation modes.
- [x] add `make` build integration
- [x] add `CMake` build integration
- [x] Rework LICENSE file to be recognizable to GitHub, and add SPDX license headers to everything
- [x] Document new features on ReadTheDocs (see https://irep.readthedocs.io/en/irep-refactor/ to preview)
- [x] Update old examples to use the new build system integration
- [x] Add github actions to test the example builds and documentation

### CODE REVIEW

File: bin/irep-generate
Comment (tomstitt) [inline bin/irep-generate]: update? `'index', 'fortran', 'lua', 'rst'`

File: wkt.mk
Comment (tomstitt) [inline wkt.mk:91]: =]

File: share/irep/irep-config.cmake
Comment (tomstitt) [inline share/irep/irep-config.cmake]: No longer a valid mode?

File: README.md
Comment (tomstitt) [inline README.md]: (minor) `...under the terms [of] the...`

**Review by tomstitt — APPROVED** [review]
Looks great! Thanks @tgamblin !

File: LICENSE
Comment (kennyweiss) [inline LICENSE]:
```suggestion
Copyright (c) 2016-2021 Lawrence Livermore National Security, LLC.
```

**Comment by tgamblin** [comment]
@tomstitt @kennyweiss: Tom's feedback should be addressed.  This is ready to merge if you want to.  I am adding docs but I can also just put in a separate PR for those if you merge this before I finish.

**Comment by rieben1** [comment]
Thanks @tgamblin, looks good.

File: README.md
Comment (kennyweiss) [inline README.md:1]: This is a great quickstart!

File: bin/l2ir
Comment (kennyweiss) [inline bin/l2ir:8]: Minor (more of a style/fyi): These comments could use lua's block comments: 

```lua
--[[
  For scalars, the type and value of the assigned value dictates the
  type and default value of the IR macro output.
  a = 3    -- is an integer named "a" with default value "3"
  b = 3.14 -- is a double, with default "3.14".
  ...
--]]

```

File: docs/build.rst
Comment (kennyweiss) [inline docs/build.rst]:
```suggestion
is, ``app``, ``liba``, and ``libb`` need to include these headers and
```

File: docs/overview.rst
Comment (kennyweiss) [inline docs/overview.rst]:
```suggestion
The Intermediate Representation2 (or IR or IREP) is a tool to simplify
```

Comment (kennyweiss) [inline docs/overview.rst]: >Representation2

Is the `2` a typo?

File: docs/quick_start.rst
Comment (kennyweiss) [inline docs/quick_start.rst:153]: These lines look weird in the generated docs.
(See: https://irep.readthedocs.io/en/irep-refactor/quick_start.html#callback-functions)

Do they need a pair of extra spaces at the end of each line to enforce line breaks?
Or a way to indicate that it is a code / verbatim environment?

File: docs/wkt/mesh/introduction.rst
Comment (kennyweiss) [inline docs/wkt/mesh/introduction.rst:4]: 👍 Thanks for the comment!

File: docs/wkt/wkt_mesh.h
Comment (kennyweiss) [inline docs/wkt/wkt_mesh.h:5]: Is there an easy way to add the generated docs for the `mesh` table to RTD as an example?

File: examples/c/GNUmakefile
Comment (kennyweiss) [inline examples/c/GNUmakefile:1]: It might be helpful to add a brief README to each of these examples to help users configure these builds.

File: examples/cxx-cmake/CMakeLists.txt
Comment (kennyweiss) [inline examples/cxx-cmake/CMakeLists.txt:1]: Note for the future: It would be interesting to look into how well `add_wkt_library` interoperates with a `blt`-based build system.

**Review by kennyweiss — APPROVED** [review]
Thanks @tgamblin -- this is great! Thanks also for the thorough documentation and CI/docs setups.

I saw a few minor typos and added some minor suggestions -- nothing that should hold up this PR though.

File: .github/workflows/linux_build_tests.yaml
Comment (kennyweiss) [inline .github/workflows/linux_build_tests.yaml]: Does this run the examples? or were you planning to add that in the future?

It seems to me that you'd have to `cd` into the `examples` directory? And it doesn't look like it is actually building/running the examples: 
https://github.com/LLNL/irep/runs/2049666979?check_suite_focus=true#step:5:4

Edit: That was supposed to point into the `Build and Run Examples step` which has the following output:
```make
Run make LUA_INCLUDE=-I/usr/include/lua5.1 LUA_LIBRARIES="-L/usr/lib/x86_64-linux-gnu -llua5.1"
  make LUA_INCLUDE=-I/usr/include/lua5.1 LUA_LIBRARIES="-L/usr/lib/x86_64-linux-gnu -llua5.1"
  shell: /bin/bash -e {0}
make: Nothing to be done for 'all'.
```

Comment (kennyweiss) [reply .github/workflows/linux_build_tests.yaml]: (I could be missing something though...)

Comment (tgamblin) [reply .github/workflows/linux_build_tests.yaml]: oh nice catch.  I think I deleted the `test` target when I added the lua libs -- lemme fix that.

File: bin/l2ir
Comment (tgamblin) [reply bin/l2ir:8]: do you like that better?

File: LICENSE
Comment (tgamblin) [reply LICENSE]: fixed.

File: docs/overview.rst
Comment (tgamblin) [reply docs/overview.rst]: Yeah that's leftover from stripping control chars from Lee's original troff file:

```
 groff  -t -e -mandoc -T ascii ./irep-intro.tr |col -bx > ../../docs/overview.rst
```

There were `4m`'s and `24m`'s left.  Must've been from my first try at replacing the `4m`'s first.

fixed.

File: docs/quick_start.rst
Comment (tgamblin) [reply docs/quick_start.rst:153]: should be `::` (for preformatted block) instead of `:` on the preceding line of text.  Fixed.

File: docs/wkt/wkt_mesh.h
Comment (tgamblin) [reply docs/wkt/wkt_mesh.h:5]: It was supposed to be there but it looks like `lua` is not installed on RTD, and you cannot install it, so the build failed. I checked it into the repo and now it should show up.

File: .github/workflows/linux_build_tests.yaml
Comment (tgamblin) [reply .github/workflows/linux_build_tests.yaml]: Ok this is fixed -- I also split it out into 3 make calls that show up as steps in the action.

File: bin/l2ir
Comment (kennyweiss) [reply bin/l2ir:8]: Not sure.  It's fine the way it is, I just wanted to point it out in case that looked cleaner.

File: examples/c/GNUmakefile
Comment (tgamblin) [reply examples/c/GNUmakefile:1]: done. I added one README in examples, as there is a lot of commonality.

File: docs/overview.rst
Comment (tgamblin) [reply docs/overview.rst]: fixed

File: docs/build.rst
Comment (tgamblin) [reply docs/build.rst]: fixed

File: bin/l2ir
Comment (tgamblin) [reply bin/l2ir:8]: looks like lua docstrings use `--` for function docs -- so I think I'll leave it.

File: README.md
Comment (tgamblin) [reply README.md:1]: thanks!

**Pull request merged**
